language/move-lang/src/naming/uses.rs

```diff
@@ -50,63 +61,109 @@ impl<'a> Context<'a> {
+) -> DiGraphMap<&ModuleIdent, ()> {
+    let edges = deps
+        .iter()
+        .flat_map(|(parent, children)| children.iter().map(move |(child, _)| (parent, child)));
+    DiGraphMap::from_edges(edges)
+}
+
+fn cycle_error(
+    deps: &BTreeMap<ModuleIdent, BTreeMap<ModuleIdent, BTreeMap<DepType, Loc>>>,
+    cycle_ident: ModuleIdent,
+) -> Error {
+    let graph = dependency_graph(deps);
+    let cycle = shortest_cycle(&graph, &cycle_ident);
+
+    let mut cycle_strings: String = cycle
+        .windows(2)
+        .map(|pair| {
+            let upstream = pair[0];
+            let downstream = pair[1];
+            let relations = deps.get(upstream).unwrap().get(downstream).unwrap();
+            let verb = if relations.contains_key(&DepType::Use) {
+                "uses"
+            } else {
+                assert!(relations.contains_key(&DepType::Friend));
+                "is a friend of"
+            };
+            format!("'{}' {} ", upstream, verb)
+        })
+        .collect();
+    cycle_strings.push_str(&format!("'{}'", cycle.last().unwrap()));
 
     // For printing uses, sort the cycle by location (earliest first)
-    let cycle_strings = cycle
-        .iter()
-        .map(|m| format!("'{}'", m))
-        .collect::<Vec<_>>()
-        .join(" uses ");
-
-    let (used_loc, user, used) = best_cycle_loc(context, cycle);
-
-    let use_msg = format!("Invalid use of module '{}' in module '{}'.", used, user);
-    let cycle_msg = format!(
-        "Using this module creates a dependency cycle: {}",
-        cycle_strings
-    );
-    vec![(used_loc, use_msg), (used_loc, cycle_msg)]
+    let (dep_type, cycle_loc, member, host) = best_cycle_loc(deps, cycle);
+
+    let (use_msg, cycle_msg) = match dep_type {
+        DepType::Use => (
+            format!("Invalid use of module '{}' in module '{}'.", member, host),
+            format!(
+                "Using this module creates a dependency cycle: {}",
+                cycle_strings
+            ),
+        ),
+        DepType::Friend => (
+            format!("Invalid friend '{}' in module '{}'", member, host),
+            format!(
+                "This friend relationship creates a dependency cycle: {}",
+                cycle_strings
+            ),
+        ),
+    };
+    vec![(cycle_loc, use_msg), (cycle_loc, cycle_msg)]
 }
 
 fn best_cycle_loc<'a>(
-    context: &'a Context,
+    deps: &'a BTreeMap<ModuleIdent, BTreeMap<ModuleIdent, BTreeMap<DepType, Loc>>>,
     cycle: Vec<&'a ModuleIdent>,
-) -> (Loc, &'a ModuleIdent, &'a ModuleIdent) {
+) -> (DepType, Loc, &'a ModuleIdent, &'a ModuleIdent) {
     let len = cycle.len();
     assert!(len >= 3);
     let first = cycle[0];
-    let user = cycle[len - 2];
-    let used = cycle[len - 1];
-    assert!(first == used);
-    let used_loc = context.neighbors.get(user).unwrap().get(used).unwrap();
-    (*used_loc, user, used)
+    let upstream = cycle[len.wrapping_sub(2)];
+    let downstream = cycle[len.wrapping_sub(1)];
```

Comment:
I would remove this wrapping sub. There is an assert right above that the length is greater than 3. (I really hate that nag that says this is unsafe, its perfectly fine and not in a critical code path)